src/core/toPath.js

```diff
@@ -102,139 +102,70 @@ const allowingArrays = fn => arg => {
-
-      // If array index is a quoted string
-      if (quoted) {
-        // Move index after the string quote
-        index++
-
-        // Look for the next unescaped matching string quote
-        let endQuoteIndex, quotedIndex = index
-        do {
-          endQuoteIndex = str.indexOf(quote, quotedIndex)
-          quotedIndex = endQuoteIndex + 1
-        } while (endQuoteIndex !== -1 && str.charAt(endQuoteIndex - 1) === '\\')
-
-        // If no end quote found, stop if end of str is reached, or continue to next iteration
-        if (endQuoteIndex === -1) {
-          if (index !== str.length) path.push(str.substring(index))
-          break
-        }
-
-        // Add the content of quotes to the path, unescaping escaped quotes
-        path.push(unescapeQuotes(str.substring(index, endQuoteIndex), quote))
-
-        // Move index after end quote
-        index = endQuoteIndex + 1
-
-        // If next character is a closing square bracket, move index after it
-        if (str.charAt(index) === ']') index++
-
-        // Stop if end of str has been reached
-        if (index === str.length) break
-
-        // If next character is a dot, move index after it (skip it)
-        if (str.charAt(index) === '.') index++
-
-      } else { // If array index is not a quoted string
-
-        // Look for the closing square bracket
-        const closingBracketIndex = str.indexOf(']', index)
-
-        // If no closing bracket found, stop if end of str is reached, or continue to next iteration
-        if (closingBracketIndex === -1) {
-          if (index !== str.length) path.push(str.substring(index))
-          break
-        }
-
-        // Fetch the content of brackets and move index after closing bracket
-        const arrayIndexValue = str.substring(index, closingBracketIndex)
-        index = closingBracketIndex + 1
-
-        // If next character is a dot, move index after it (skip it)
-        if (str.charAt(index) === '.') index++
-
-        // Shorthand: if array index is the whole slice add it to path
-        if (arrayIndexValue === ':') {
-          path.push([undefined, undefined])
-        } else {
-
-          // Look for a slice quote
-          const sliceDelimIndex = arrayIndexValue.indexOf(':')
-
-          // If no slice quote found
-          if (sliceDelimIndex === -1) {
-            // Parse array index as a number
-            const nArrayIndexValue = Number(arrayIndexValue)
-
-            // Add array index to path, either as a valid index (positive int), or as a string
-            path.push(isIndex(nArrayIndexValue) ? nArrayIndexValue : arrayIndexValue)
+  const path = stringToPath2(str)
+  return str[0] === '.' ? ['', ...path] : path
+}
 
-          } else { // If a slice quote is found
+const splitAtFirstOccurence = (str, separators) => {
+  const partitionIndex = separators
+    .map(separator => str.indexOf(separator))
+    .map(index => index >= 0 ? index : str.length)
```

Comment:
You could replace this by a filter and add `str.length` in second parameter to reduce

Reply:
Ah, yes, nice idea!